Fst windowing

tomwhite opened the pull request:
This is a proposal for #229, and is also related to #281.

* Adds a new `window` function that adds windowing information to a dataset (a `windows` dimension). This just adds fixed-size windows at the moment, and doesn't take contigs into account - both could be improved in the future.
* Adds a `window_statistic` function (similar scikit-allel's `moving_statistic`) that computes a statistic for each window (using Dask's `map_overlap` function) for use by method functions (i.e. this is not a user-facing function).
* Updates `diversity`, `divergence`, and `Fst` to use windows if present.

Sample usage:

```python
>>> ds = window(ds, size=25, step=25)
>>> fst_ds = Fst(ds, estimator="Nei")
>>> fst_ds
<xarray.Dataset>
Dimensions:              (alleles: 2, cohorts: 2, cohorts_0: 2, cohorts_1: 2, ploidy: 1, samples: 10, variants: 98, windows: 4)
Coordinates:
  * cohorts_0            (cohorts_0) object 'co_0' 'co_1'
  * cohorts_1            (cohorts_1) object 'co_0' 'co_1'
Dimensions without coordinates: alleles, cohorts, ploidy, samples, variants, windows
Data variables:
    stat_Fst             (windows, cohorts_0, cohorts_1) float64 dask.array<chunksize=(4, 2, 2), meta=np.ndarray>
    cohort_allele_count  (variants, cohorts, alleles) int64 dask.array<chunksize=(98, 2, 2), meta=np.ndarray>
    call_allele_count    (variants, samples, alleles) uint8 dask.array<chunksize=(98, 10, 2), meta=np.ndarray>
    window_start         (windows) int64 0 25 50 75
    window_stop          (windows) int64 25 50 75 98
    variant_contig       (variants) int64 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0
    variant_position     (variants) int64 0 2 6 8 12 19 ... 193 193 194 196 197
    variant_allele       (variants, alleles) |S1 b'0' b'1' b'0' ... b'0' b'1'
    sample_id            (samples) <U5 'tsk_0' 'tsk_1' ... 'tsk_8' 'tsk_9'
    call_genotype        (variants, samples, ploidy) int8 0 0 0 1 1 ... 1 0 0 1
    call_genotype_mask   (variants, samples, ploidy) bool False False ... False
    sample_cohort        (samples) int32 0 0 0 0 0 1 1 1 1 1
>>> fst = fst_ds["stat_Fst"].values
>>> fst
[[[       nan 0.04802022]
  [0.04802022        nan]]

 [[       nan 0.05591201]
  [0.05591201        nan]]

 [[       nan 0.03881701]
  [0.03881701        nan]]

 [[       nan 0.02708124]
  [0.02708124        nan]]]
```

Note the `window_start` and `window_stop` variables of dimension `windows`, and the `stat_Fst` variable whose first dimension is `windows`.

tomwhite commented:
BTW this is based on #302, which should be merged first.

alimanfoo commented:
Hi @tomwhite, one small thought is that I'll often try computing some statistic with a variety of different window sizes. So we might need some way to specify a name or prefix for each set of windows being added to the same dataset.

tomwhite commented:
@alimanfoo that's good to know. Another way of achieving this would be to set `merge=False` to get a dataset for each set of windows.

ravwojdyla commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+        new_chunks = (tuple(windows_per_chunk),)
+    else:
+        # depth is 0 except in first axis
+        depth = tuple([depth] + ([0] * (values.ndim - 1)))
```
I wonder if `map_overlap` would work in this case with `depth = {0: depth}` (without the need to specify the depth on the other dimensions) 🤔

eric-czech replied:
That should work: https://github.com/dask/dask/blob/master/dask/array/overlap.py#L732 is the relevant bit for that (all others default to 0).

tomwhite replied:
Thanks for checking - fixed

ravwojdyla commented on sgkit/tests/test_window.py:
```diff
@@ -0,0 +1,46 @@
+    values_sa = arr
+    stat_sa = allel.moving_statistic(values_sa, sum_cols, size=size, step=step)
+
+    np.testing.assert_equal(stat, stat_sa)
```
Would it make sense to add more tests here, there are some interesting functions in the `window` package, like `_get_chunked_windows`? Plus test for corner cases like empty arrays etc. And maybe also a test for what happens when you window multiple times with `merge=True`. And maybe a test case for https://github.com/pystatgen/sgkit/pull/303#issuecomment-705682385

tomwhite replied:
Added tests for various window functions in https://github.com/pystatgen/sgkit/pull/303/commits/d57aefdb889fd9190b44d4439ba75e91ac8201b9

ravwojdyla commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+# Window definition (user code)
+
+
+def window(
```
Thinking about the API, I wonder if we should have a simple function to "remove" the windowing from a dataset? Otherwise we might expect people to pop `window_start`/`window_stop` vars?

tomwhite replied:
Yes, that would be useful. Opened #333

ravwojdyla reviewed (changes requested):
I like the idea of storing the window information in the `Dataset`!

tomwhite commented:
Thanks for the review @ravwojdyla! I will fix your suggestions given that the overall approach seems to be OK.

eric-czech commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+
+def _get_windows(start: int, stop: int, size: int, step: int) -> Any:
+    # Find the indexes for the start positions of all windows
+    # TODO: take contigs into account
```
Worth elevating as an issue?  I'm sure it will come up again in the context of LD estimation.

tomwhite replied:
Opened #335

eric-czech commented on sgkit/tests/test_window.py:
```diff
@@ -0,0 +1,46 @@
+    stat_sa = allel.moving_statistic(values_sa, np.sum, size=size, step=step)
+
+    np.testing.assert_equal(stat, stat_sa)
+
```
I think a test for dtype preservation when provided float32 inputs would be good for these kinds of functions.

tomwhite replied:
I added an assert to check for dtype preservation here.

tomwhite commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def _get_windows(start: int, stop: int, size: int, step: int) -> Any:
```
Changed to `ArrayType`.

tomwhite commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+    )
+
+
+def window_statistic(
```
This was from when I was prototyping, and I forgot to fix them before submitting. I have now eliminated them all except `kwargs` and `block_info`, which are harder to provide hints for.

tomwhite commented on sgkit/stats/popgen.py:
```diff
@@ -68,15 +69,33 @@ def diversity(
+        div = window_statistic(
+            pi,
+            np.sum,
+            ds.window_start.values,
+            ds.window_stop.values,
```
No, not at the moment at least.

eric-czech replied:
Should it be?  The loose guideline I have in my head is that anything that doesn't run interactively (<1s or so) across a whole genome dataset (~100M variants) should probably be lazy.  Is it worth adding an issue to make this lazy or test that the latter is true?

tomwhite replied:
Opened #340

tomwhite commented on sgkit/tests/test_popgen.py:
```diff
@@ -49,11 +58,40 @@ def test_diversity(size, chunks):
     np.testing.assert_allclose(div, ts_div)
 
 
+@pytest.mark.parametrize("size", [10])
```
Good point - especially as we now have sample size and window size in the same test.

tomwhite commented on sgkit/tests/test_popgen.py:
```diff
@@ -49,11 +58,40 @@ def test_diversity(size, chunks):
+    positions = [variant.site.position for variant in ts.variants()]
+    windows = [0] + positions[::25][1:] + [ts.sequence_length]
+    ts_div = ts.diversity(windows=windows, span_normalise=False)
+    np.testing.assert_allclose(div, ts_div)
```
Good idea. I opened #332

tomwhite commented on sgkit/tests/test_popgen.py:
```diff
@@ -86,20 +124,64 @@ def test_divergence(size, n_cohorts, chunks):
+    mpd = allel.mean_pairwise_difference_between(ac1, ac2, fill=0)
+    ska_div = allel.moving_statistic(mpd, np.sum, size=25, step=25)  # noqa: F841
+    # TODO: investigate why numbers are different
+    # np.testing.assert_allclose(div[:-1], ska_div)  # scikit-allel has final window missing
```
Done

tomwhite commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+    statistic: Callable[..., Any],
+    size: int,
+    step: int,
+    dtype: int,
```
Done

tomwhite commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+    step: int,
+    merge: bool = True,
+) -> Dataset:
+    """Add windowing information to a dataset."""
```
I expanded this, and added new variables.

tomwhite commented on sgkit/window.py:
```diff
@@ -0,0 +1,174 @@
+    if depth > values.chunks[0][-1]:
+        chunk0 = values.chunks[0]
+        new_chunk0 = tuple(list(chunk0[:-2]) + [chunk0[-2] + chunk0[-1]])
+        # None means don't rechunk along that axis
+        new_chunks = tuple(list([new_chunk0] + ([None] * (len(chunk0) - 1))))  # type: ignore
+        values = values.rechunk(new_chunks)
```
Thanks - done

tomwhite commented:
I've addressed all the feedback, except for the comment about adding more tests for the window module. I'll try to write some more, but otherwise this is probably ready to be merged.

jeromekelleher commented on sgkit/window.py:
```diff
@@ -0,0 +1,206 @@
+    ----------
+    ds
+        Genotype call dataset.
+    size
```
Hmm, I'm not so sure about this API, it feels a bit inflexible. People are going to want to window and plot in terms of physical distance and not just in terms of the number of variants (see the proposal in #229). I see that using the number of variants aligns much more nicely with chunking. Is this a fundamental restriction - will we always need windows defined in this way to have equal numbers of variants? If we did want to do things like "compute Fst in 100Kb windows", is this incompatible with the way we're doing things in Dask?

It's not obvious what the utility of the ``step`` parameter is either - I'd imagine the vast majority of people will want contiguous windows (unless I'm misunderstanding what this does?). It should at least have a default such that users do get contiguous windows and don't need to specify this value all the time.

jeromekelleher replied:
I really like this idea @tomwhite - how about we create a follow up issue for this so, and get this much merged?

tomwhite replied:
Filed #341

jeromekelleher reviewed (commented):
This looks good, but I worry that we're painting ourselves into a bit of a corner with the names of things here, given that this is so tied to variant indexes. How will we deal with windowing in physical distance? I think we should have a good idea how this is going to work and how we're going to name things before finalising this API.

tomwhite commented on sgkit/tests/test_popgen.py:
```diff
@@ -39,28 +48,57 @@ def ts_to_dataset(ts, chunks=None, samples=None):
+
+    # Calculate diversity using tskit windows
+    # Find the variant positions so we can have windows with a fixed number of variants
+    positions = [variant.site.position for variant in ts.variants()]
```
Thanks for the suggestion - I've used `ts.tables.sites.position` instead

jeromekelleher commented:
@ravwojdyla - have the changes you requested been made?

jeromekelleher commented:
Looks great @tomwhite - please add the auto-merge flag when you're happy to merge this (I guess with an issue flagging the current limitations as discussed above)
